File: SatinalmaPro.Shared/Helpers/BildirimTekillestirme.cs
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Helpers;

/// <summary>Aynı talep+tip+hedef için yinelenen bildirimleri birleştirir.</summary>
public static class BildirimTekillestirme
{
    public static List<BildirimKaydi> Tekille(IEnumerable<BildirimKaydi> kaynak)
    {
        return kaynak
            .GroupBy(BildirimMantikAnahtari.Olustur)
            .Select(BirlestirGrup)
            .OrderByDescending(b => b.GuncellemeUtc)
            .ThenByDescending(b => b.OlusturmaTarihi, StringComparer.Ordinal)
            .ToList();
    }

    private static BildirimKaydi BirlestirGrup(IEnumerable<BildirimKaydi> grup)
    {
        var sirali = grup
            .OrderByDescending(b => b.Okundu)
            .ThenByDescending(b => b.GuncellemeUtc)
            .ToList();

        var birincil = Kopyala(sirali[0]);
        foreach (var diger in sirali.Skip(1))
        {
            birincil.Okundu = birincil.Okundu || diger.Okundu;
            birincil.GuncellemeUtc = Math.Max(birincil.GuncellemeUtc, diger.GuncellemeUtc);

            if (string.IsNullOrWhiteSpace(birincil.InboxDocId) && !string.IsNullOrWhiteSpace(diger.InboxDocId))
                birincil.InboxDocId = diger.InboxDocId;
            if (string.IsNullOrWhiteSpace(birincil.Baslik))
                birincil.Baslik = diger.Baslik;
            if (string.IsNullOrWhiteSpace(birincil.Mesaj))
                birincil.Mesaj = diger.Mesaj;
        }

        return birincil;
    }

    private static BildirimKaydi Kopyala(BildirimKaydi kaynak) => new()
    {
        Id = kaynak.Id,
        Baslik = kaynak.Baslik,
        Mesaj = kaynak.Mesaj,
        Tip = kaynak.Tip,
        TalepId = kaynak.TalepId,
        HedefRol = kaynak.HedefRol,
        HedefUid = kaynak.HedefUid,
        OlusturanUid = kaynak.OlusturanUid,
        OlusturanAd = kaynak.OlusturanAd,
        OlusturmaTarihi = kaynak.OlusturmaTarihi,
        Okundu = kaynak.Okundu,
        GuncellemeUtc = kaynak.GuncellemeUtc,
        InboxDocId = kaynak.InboxDocId,
        DeepLink = kaynak.DeepLink,
        EventCode = kaynak.EventCode,
        DesktopRoute = kaynak.DesktopRoute
    };
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/android/core/helpers/BildirimTekillestirme.kt
package com.satinalmapro.android.core.helpers

import com.satinalmapro.android.core.model.BildirimRecord
import com.satinalmapro.android.core.model.UserProfile

/** Aynı talep+tip+hedef için yinelenen bildirimleri birleştirir. */
object BildirimTekillestirme {
    fun tekille(kaynak: Collection<BildirimRecord>): List<BildirimRecord> =
        kaynak
            .groupBy { BildirimMantikAnahtari.olustur(it) }
            .values
            .map { birlestirGrup(it) }
            .sortedWith(
                compareByDescending<BildirimRecord> { it.guncellemeUtc }
                    .thenByDescending { it.olusturmaTarihi }
            )

    fun inboxIleBirlestir(
        legacy: List<BildirimRecord>,
        inbox: List<BildirimRecord>,
        user: UserProfile?,
        kullaniciyaMi: (BildirimRecord, UserProfile?) -> Boolean
    ): List<BildirimRecord> {
        val sonuc = inbox.toMutableList()
        for (l in legacy) {
            if (inbox.any { BildirimMantikAnahtari.olustur(it) == BildirimMantikAnahtari.olustur(l) }) continue
            if (user != null && !kullaniciyaMi(l, user)) continue
            sonuc.add(l)
        }
        return tekille(sonuc)
    }

    private fun birlestirGrup(grup: Collection<BildirimRecord>): BildirimRecord {
        val sirali = grup.sortedWith(
            compareByDescending<BildirimRecord> { it.guncellemeUtc }
                .thenByDescending { it.olusturmaTarihi }
        )
        var birincil = sirali.first()
        sirali.drop(1).forEach { diger ->
            birincil = birincil.copy(
                okundu = birincil.okundu || diger.okundu,
                guncellemeUtc = maxOf(birincil.guncellemeUtc, diger.guncellemeUtc),
                inboxDocId = birincil.inboxDocId?.takeIf { it.isNotBlank() } ?: diger.inboxDocId,
                baslik = birincil.baslik.ifBlank { diger.baslik },
                mesaj = birincil.mesaj.ifBlank { diger.mesaj }
            )
        }
        return birincil
    }
}
